### PR TITLE
feat: add modified vectorx contracts

### PR DESCRIPTION
**Pull request by xavierdmello**
modify existing vectorx contracts to work with the sp1 prover and sp1-vectorx program

### CODE REVIEW

File: .gitmodules
Comment (ratankaliani) [inline .gitmodules:1]: You should delete /lib/sp1-contracts, as it's unused.

File: contracts/src/VectorX.sol [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+/// storage corruption during upgrades.
+contract VectorX is IVectorX, TimelockedUpgradeable {
+    /// @notice Indicator of if the contract is frozen.
+    bool public frozen;
+
+    /// @notice The latest block that has been committed.
+    uint32 public latestBlock;
+
+    /// @notice The latest authority set id used in commitHeaderRange.
+    uint64 public latestAuthoritySetId;
```
Comment (ratankaliani) [inline contracts/src/VectorX.sol:25]: Seems like you deleted `headerRangeFunctionID` and `rotateFunctionId`. Please add these back to ensure backwards compatibility. Mark them as `_deprecated`, along with any other storage variables you may have deprecated.

https://forum.openzeppelin.com/t/how-to-handle-contract-unused-variables-when-using-upgradeable-smart-contract/36936

File: contracts/src/VectorX.sol [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+        }
+
+        bytes32 trustedHeader = blockHeightToHeaderHash[latestBlock];
+        if (trustedHeader == bytes32(0)) {
+            revert TrustedHeaderNotFound();
+        }
+        bytes32 authoritySetHash = authoritySetIdToHash[_authoritySetId];
+        if (authoritySetHash == bytes32(0)) {
+            revert AuthoritySetNotFound();
+        }
```
Comment (ratankaliani) [inline contracts/src/VectorX.sol]: Add additional checks here which confirm that the contract's `trustedHeaderHash` and `authoritySetHash` match those in the `publicValues` of the proof. tl;dr `trustedHeaderHash` = `contractTrustedHeaderHash` and same for the authority set hash.

File: contracts/src/VectorX.sol [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+    /// @param proof The proof bytes
+    /// @param publicValues The public commitments from the proof
+    function rotate(bytes calldata proof, bytes calldata publicValues) external {
+        if (frozen) {
+            revert ContractFrozen();
+        }
+
+        (uint8 proofTypeInt, , bytes rotateOutputs) = abi.decode(publicValues, (uint8, bytes, bytes));
+        ProofType proofType = ProofType(proofTypeInt);
+        (uint64 _currentAuthoritySetId, bytes32 newAuthoritySetHash) = abi.decode(rotateOutputs, (uint64, bytes32));
```
Comment (ratankaliani) [inline contracts/src/VectorX.sol]: You should add the `currentAuthoritySetHash` to the outputs of the circuit, and then in this contract function confirm that it equals the contract's authority set hash corresponding to that authority set id.

File: primitives/src/types.rs [hunk 4]
```diff
@@ -14,11 +14,17 @@ pub type HeaderRangeOutputs = sol! {
     tuple(uint32, bytes32, uint64, bytes32, uint32, bytes32, bytes32)
 };
```
Comment (ratankaliani) [inline primitives/src/types.rs]: Please add `target_header_hash` to the outputs here. 

Any value with `evm_write` or `evm_read` should be included in the outputs:

https://github.com/succinctlabs/vectorx/blob/95b712c220d6d7cabcc5750b00d88d7505cbf563/circuits/rotate.rs#L85-L108

https://github.com/succinctlabs/vectorx/blob/95b712c220d6d7cabcc5750b00d88d7505cbf563/circuits/header_range.rs#L31-L59

**Review by ratankaliani — COMMENTED** [review]
Left some comments, there's some missing values from the outputs. Also added some comments for the changes to the contracts.

File: .gitmodules
Comment (xavierdmello) [reply .gitmodules:1]: I'm using ISP1Verifier.sol, should I delete the folder and copy the file to src/interfaces?

File: contracts/.vscode/settings.json
Comment (ratankaliani) [inline contracts/.vscode/settings.json:1]: This file can be deleted.

File: contracts/README.md
Comment (ratankaliani) [inline contracts/README.md:1]: Remove if unused.

File: contracts/src/VectorX.sol [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+import {ISP1Verifier} from "@sp1-contracts/ISP1Verifier.sol";
+
+/// @notice VectorX is a light client for Avail's consensus.
+/// @dev The light client tracks both the state of Avail's Grandpa consensus and Vector, Avail's
+///     data commitment solution.
+/// @dev Ensure that all new storage variables are placed after existing storage variables to avoid
+/// storage corruption during upgrades.
+contract VectorX is IVectorX, TimelockedUpgradeable {
+    /// @notice Indicator of if the contract is frozen.
+    bool public frozen;
```
Comment (ratankaliani) [inline contracts/src/VectorX.sol:15]: Looks like you deleted `address gateway` here.

https://github.com/succinctlabs/vectorx/blob/main/contracts/src/VectorX.sol#L18

File: contracts/src/VectorX.sol [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+        uint32 headerRangeCommitmentTreeSize;
+        bytes32 vectorXProgramVkey;
+        address verifier;
+    }
+
+    /// @notice The verification key for the VectorX program.
+    bytes32 public vectorXProgramVkey;
+
+    /// @notice The deployed SP1 verifier contract.
+    ISP1Verifier public verifier;
```
Comment (ratankaliani) [inline contracts/src/VectorX.sol]: cc @mattstam Can probably move these above the struct?

File: primitives/src/consts.rs [hunk 7]
```diff
@@ -11,4 +11,10 @@ pub const WEIGHT_LENGTH: usize = 8;
...
 
 // ABI-encoded length of the header range outputs.
-pub const HEADER_OUTPUTS_LENGTH: usize = 32 * 7;
\ No newline at end of file
+pub const HEADER_OUTPUTS_LENGTH: usize = 32 * 8;
+
+// ABI-encoded length of the rotate outputs.
+pub const ROTATE_OUTPUTS_LENGTH: usize = 32 * 3;
+
+// ABI-encoded length of final proof output.
```
Comment (ratankaliani) [inline primitives/src/consts.rs:19]: Where is this used? Don't see it anywhere in this diff which indicates it's either unused, or `sp1-vectorx` doesn't build.

File: script/src/main.rs
Comment (ratankaliani) [inline script/src/main.rs:1]: Given that you want to test both `headerRange` and `rotate`, it's probably best to turn this into a helper function that accepts the proof type enum and then the inputs as options, and unwraps them if they're necessary for the proof you want to generate. You can do this in your next PR.

File: primitives/src/consts.rs (hunk 7 above)
Comment (xavierdmello) [reply primitives/src/consts.rs:19]: its used in program/src/main.rs, I think I might have made a weird commit bc foundry messes with your git

File: contracts/src/VectorX.sol [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+
+    /// @notice Adds the authority set hash for the next authority set id.
+    /// @param proof The proof bytes
+    /// @param publicValues The public commitments from the proof
+    function rotate(bytes calldata proof, bytes calldata publicValues) external {
+        if (frozen) {
+            revert ContractFrozen();
+        }
+
+        (uint8 proofTypeInt,, bytes memory rotateOutputs) = abi.decode(publicValues, (uint8, bytes, bytes));
```
Comment (ratankaliani) [inline contracts/src/VectorX.sol:249]: nit: Why are there two comma's here? Are you sure this compiles?

**Review by ratankaliani — APPROVED** [review]
Generally looks good, make sure that the contracts build before merging.

File: contracts/src/VectorX.sol (hunk 8 above)
Comment (xavierdmello) [reply contracts/src/VectorX.sol:249]: solidity syntax for throwing away a return value (similar to rust underscore)

**Pull request merged**